_posts/2019-03-05-build-firefox-extension-with-ringcentral-embeddable.md

```diff
@@ -0,0 +1,304 @@
+  if (request.type === 'rc-register-service') {
+    this.registerService(sendResponse)
+    return true;
+  }
+  if (request.type === 'rc-post-message-request') {
+    if (request.path === '/authorize') {
+      this.onAuthorize(request.body.authorized);
+      sendResponse({ data: 'ok' });
+    }
+    if (request.path === '/contacts') {
+      this.onGetContacts(request, sendResponse);
+    }
+    if (request.path === '/conference/invite') {
+      this.createCalendarEvent(request, sendResponse);
+    }
+    if (request.path === '/activities') {
+      this.getContactGmails(request, sendResponse);
+    }
+    if (request.path === '/activity') {
+      this.openGmailPage(request, sendResponse);
+    }
+    if (request.path === '/contacts/search') {
+      this.onContactSearch(request, sendResponse);
+    }
+    if (request.path === '/contacts/match') {
+      this.onContactMatch(request, sendResponse);
+    }
+    return true;
+  }
+});
+```
+
+### Integrate Google related feature
+
+After injecting RingCentral Embeddable widget into webpage, we also want to integrate third party service into RingCentral Embeddable. In this section, we will integration Google Authorization, Google Contacts, Google Calendar and Gmail into widget.
+
+![ringcentral embeddable integrate with gsuit](https://user-images.githubusercontent.com/7036536/55455660-0e8f0000-5617-11e9-89be-17c008c7d3f4.png)
+
+### Integrate Google authorization into widget:
+
+Add following codes into `background.js`
+
+```js
+chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
+  // Register serivce into widget
+  if (request.type === 'rc-register-service') {
+    registerService(sendResponse)
+    return true;
+  }
+  if (request.type === 'rc-post-message-request') {
+    // Response to authorize request
+    if (request.path === '/authorize') {
+      this.onAuthorize(request.body.authorized);
+      sendResponse({ data: 'ok' });
+    }
+  }
+});
+
+async function registerService(sendResponse) {
+  const authorized = await googleClient.checkAuthorize();
+  // Register Google service
+  sendResponse({
+    action: 'registerService',
+    service: {
+      name: 'Google',
+      authorizationPath: '/authorize',
+      authorizedTitle: 'Unauthorize',
+      unauthorizedTitle: 'Authorize',
+      authorized,
+    }
+  });
+}
+
+async function onAuthorize(authorized) {
+  if (!authorized) {
+    await googleClient.authorize();
+  } else {
+    await googleClient.unAuthorize();
+  }
+  const newAuthorized = await googleClient.checkAuthorize();
+  if (newAuthorized) {
+    googleClient.setUserInfo();
+  }
+  await sendMessageToContentAndStandalong(
+    { action: 'authorizeStatusChanged', authorized: newAuthorized }
+  );
+}
+```
+
+You can get full google client codes in here. Firstly, we register authorize path when we register service. Then we build a message listener to response authorize request.
```

Comment:
full google client codes => `the` full google client `code`
Firstly, we register `a authorization path` when we register `the service`. Then we build a message listener to `respond to the authorization` request.